### PR TITLE
Fix potential memory leak 

### PR DESCRIPTION
**Pull request by iskander232**
Fix memory leak in case of exiting the rope_insert function with an error and add test for it.

### CODE REVIEW

**Review by Korablev77 — COMMENTED** [review]
General comments: please squash your commits so we have one patch per bug fix (until they have different goals); write meaningful commit message - declare the problem, why it takes place and how have you fixed it (even in case of such simple fix).
Before sending patch to the review, please make self-review: I see several codestyle violations (both in code and in test).
Otherwise great job! It would be nice to add comments to the test as well.

File: src/lib/salad/rope.h [hunk 1]
```diff
@@ -374,8 +374,10 @@ rope_insert(struct rope *rope, rope_size_t offset, rope_data_t data,
 		if (offset != 0) {
 			struct rope_node *split_node;
 			split_node = rope_node_split(rope, **p_end, offset);
-			if (split_node == NULL)
-				return -1;
+			if (split_node == NULL){
```
Comment (Korablev77) [inline src/lib/salad/rope.h:377]: Nit: missing space after '('

File: test/unit/rope_out_of_memory.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
+#include "unit.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:4]: Including stdio and stdlib headers seem to be redundant

File: test/unit/rope_out_of_memory.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+/** Counting the number of allocated nodes */
+static size_t allocated_count;
+/** Counting the number of freed nodes */
+static size_t freed_count;
+/** Number of remaining node allocationы */
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:10]: ы at the end of sentence :)

File: test/unit/rope_out_of_memory.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#define ROPE_ALLOC_F mem_alloc_sanitizer
+#define ROPE_FREE_F mem_free_sanitizer
+#define ROPE_SPLIT_F str_getn
+#define rope_data_t char *
+#define rope_ctx_t void *
+
+#include "salad/rope.h"
+
+static inline struct rope *
+prepare_test(int max_alloc_count){
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:49]: '{' should be at new line

File: test/unit/rope_out_of_memory.c [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Test for bug fix #5788.
+ * Try to exit from rope_insert due to out of memory
+ * by limiting the available node allocations and
+ * check that the number of allocated
+ * nodes and the number of freed nodes are the same.
+ */
+static void
+out_of_memory_stress_test(){
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:64]: '{' should be at new line

File: test/unit/rope_out_of_memory.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        header();
+
+        const int iterations = 500;
+        plan(iterations);
+        for(int max_allocs = 1; max_allocs <+ iterations; ++max_allocs){
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:69]: Spaces after 'for' word and after '(' are missing
<+ ?

File: test/unit/rope_out_of_memory.result [hunk 7]
```diff
@@ -0,0 +1,503 @@
+	*** out_of_memory_stress_test ***
+1..500
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.result:2]: Could you please reduce result output? 500 checks is too much imho

File: test/unit/rope_out_of_memory.c [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        check_plan();
+        footer();
+}
+
+int
+main()
+{
+        out_of_memory_stress_test();
+        return 0;
+}
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:87]: No new line at the end of file

File: test/unit/rope_out_of_memory.c [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+ * Try to exit from rope_insert due to out of memory
+ * by limiting the available node allocations and
+ * check that the number of allocated
+ * nodes and the number of freed nodes are the same.
+ */
+static void
+out_of_memory_stress_test(){
+        header();
+
+        const int iterations = 500;
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.c:67]: I'd reduce the number of iterations. Anyway it starts to fail after several invocations of rope_insert..

File: test/unit/rope_out_of_memory.result [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+ok 462 - all allocated nodes freed
+ok 463 - all allocated nodes freed
+ok 464 - all allocated nodes freed
+ok 465 - all allocated nodes freed
+ok 466 - all allocated nodes freed
+ok 467 - all allocated nodes freed
+ok 468 - all allocated nodes freed
+ok 469 - all allocated nodes freed
+ok 470 - all allocated nodes freed
+# Looks like you planned 500 tests but ran 499.
```
Comment (Korablev77) [inline test/unit/rope_out_of_memory.result:473]: Looks wrong

**Comment by Korablev77** [comment]
If it is regression test as a rule we add gh-xxxx prefix to it. So personally I'd call this test 'gh-5788-rope-insert-oom'

File: src/lib/salad/rope.h (hunk 1 above)
Comment (iskander232) [reply src/lib/salad/rope.h]: fix

File: test/unit/rope_out_of_memory.c (hunk 2 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.c (hunk 3 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.c (hunk 4 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.c (hunk 5 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.c (hunk 6 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.result (hunk 7 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.result]: reduced to 20

File: test/unit/rope_out_of_memory.c (hunk 8 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: fix

File: test/unit/rope_out_of_memory.c (hunk 9 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.c]: reduced to 20

File: test/unit/rope_out_of_memory.result (hunk 10 above)
Comment (iskander232) [reply test/unit/rope_out_of_memory.result]: fix

**Review by Korablev77 — APPROVED** [review]
Thx, great work. LGTM

**Review by alyapunov — APPROVED** [review]
LGTM

**Comment by Korablev77** [comment]
Pushed to master, 2.7, 2.6 and 1.10.

**Pull request closed without merging**